docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,140 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability to extend the DNS configuration of the gateway with the DNSPolicy resource. This resource allows 
+users to configure health checks. As a result of configuring health checks, the controller creates the health checks in 
+Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it 
+becomes unhealthy, and enabling it again when it becomes healthy again.
+
+This feature has a few shortfalls:
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, 
+or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+3. requires using the gateway controller to implement, maintain and test multiple providers
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+
+### Nongoals
+* Ability for the health checks to reach endpoints in private networks
+* Transparently keep support for other health check providers like Route53
+
```

Comment:
I think adding some use cases here would be good also:

```
As a gateway administrator, I would like to define a health check that each service sitting behind a particular listener across the production clusters has to implement to ensure we can automatically respond, failover and mitigate a failing instance of the service
```
@david-martin ^ FYI

Reply:
I think that's a reasonable thing a gateway administrator would like to have.

In reality, I wonder if there's a different use case that's needed as well?
Will our users be building new apps (or think it reasonable to extend them all to conform to a health check API)?
Or will there be a lot of 'lift and shift' of existing apps. Perhaps there are existing apps that are not moving, but instead are being routed to via vpns, submariner, skupper etc.. from a central gateway.
In these cases, would it make more sense to allow per listener/host health checks to be defined based on the app?

Reply:
I think allowing policy to target listeners is something we will want. This is not something that policy attachment supports in its targeting yet though. I think this is something we should discuss with them using use cases like this. Something else I have being given thought to is whether we could allow a policy to be defined in the spoke that is leveraged by the control plane (IE health set on the HTTPRoute rather than the gateway for some use cases) This feels  complex though and is something we would want to give careful thought. Esp when it comes to making sense of what is happening

Reply:
Does direct policy attachment dictate that you can only have one policy attached to a target? I know the kuadrant policies do this, but is that a requirement? I was wondering if we could have multiple polices attached to a single gateway (direct attachment), but specify what hosts it should be applied to in a similar way to the RateLimit and AuthConfig policies https://github.com/Kuadrant/kuadrant-operator/blob/main/api/v1beta1/ratelimitpolicy_types.go#L103 and https://github.com/Kuadrant/kuadrant-operator/blob/main/controllers/ratelimitpolicy_controller.go#L161. We don't do any of this in the DNSPolicy currently and just accept a policy will act against any listener, but it could be scoped. Problem with that and multiple polices would be you'd only want one policy acting on any given listener.

Reply:
> would it make more sense to allow per listener/host health checks to be defined based on the app

I don't this host / listener specific details belong in a gateway policy. I have always considered policies to be the abstract definition of some rule that can be applied to multiple gateways. I feel that specifying particular hosts in a policy contradicts that philosophy. 

i.e. a policy with specifically defined hosts is only really useful to one particular gateway, which undermines the purpose of it being a policy at all.

Reply:
> I don't think host / listener specific details belong in a gateway policy.

While I don't disagree with this (the scope of a policy targeting a gateway), I am questioning the usefulness of defining a health check that applies to all listeners vs. the ability to define a health check per listener/host/application.

Reply:
I see, this concept is like a... health check for a specific route or endpoint, which I suppose would be a different type of object to a DNSHealthCheckPolicy. I see value in that alright, but it probably warrants it's own issue / proposal separate to this one?

Reply:
Yes different issue, @david-martin what is the value of having to define a health check for each listener vs all listeners in a gateway?
Where I could potentially see it being useful is 
1) if the applications are in a multi-tenanted environment and don't all have to conform to a given "production standard" with the current design the check at the gateway level could still be useful to those that want it. Adding 404 as an allowed response code 
2) an extension to the above, if the apps all have different auth they want to apply

I think the design here is focused around an single org where a "ops/sre" team are managing the environment and so want to specify a health check for all listeners in that environment. Thinking about allowing this type of policy at the data plane level, is worth exploration though. Just as is allowing RLP and AuthPolicy at the MC CP level.

If you're in a agreement in the value of this health check design lets follow this up with additional issues ?

Reply:
@roivaz interested in your thoughts here from an operational point of view

Reply:
> If you're in a agreement

I'd love to have this decision be more data driven.
Based on my own previous experiences, individual applications tend to be developed by different teams/people at different stages of the overall group of applications evolution. Different patterns may be necessary or evolve over time as well with regards endpoints (e.g. REST vs graphql vs custom). I do like the story of conforming to the same set of endpoints across applications with different hostnames for dns health checks.
In reality, how often is that possible?

Perhaps this could be similar to other policies like Auth where it can be defined at a coarse grain in the control plane, and be more fine grained in the data plane when it isn't specific enough. For example, I'd enforce that all applications/listeners/hosts should give a 200 OK on `/` for them to be healthy. If that needs to be changed, that's something configurable per application/listener/host... somewhere.

Reply:
Yeah I think what you are describing is a potential evolution that we should consider. We know the use case exits for DNS level health checks as it is offered by all the main providers. So it is that data point that drives the current design.

Reply:
I agree with @david-martin  that very rarely a wide group of applications conform to the same pattern regarding health checks, even though in kubernetes the /healthz and /readyz have become very popular, many applications/orgs still use custom patterns and those can vary between teams.

Also, things like timeouts or retries should be configurable per listener, because load might be different across different services, and different services might respond very differently in terms of latency, even orders of magnitude different. 

I think an organization would find it very hard to have a set of default values that can apply to every listener in a gateway unless all services exposed through it are very similar. This would end up forcing users to group similar services in the same gateway and deploying other gateways for those that don't follow the pattern or behave very differently on terms of latency.

Reply:
ok so I think here we are also expanding the feature set? So there are some potential follow ons:

1)  allow per listener health check to be defined (this is not supported by policy attachment currently) we would need to do some form of custom targeting 
2) potentially allow overriding the health check at the HTTPRoute level
3) consider additional policies that cover things like timeout and retry

What I would say is while limited in its current design nothing we are doing prevents us adding this type of functionality afterwards. 
IE (new targeting mechanism is back compatible as if not specified default is all listeners)
introducing support to override the health check lower down the hierarchy 
etc.
For now having some form of simple health check support seems reasonable. If you don't want to enforce it you can set 404 as an acceptable response? WDYT

Reply:
I think so.

> If you don't want to enforce it you can set 404 as an acceptable response?

I see what you're getting at, and that gets around the problem.
However, it should also possible to completely disable healthchecks for a gateway (maybe that's already possible in the spec by not setting it?)
If it can be disabled, it shouldn't get in the way of users that want to wait for the more flexible functionality per listener.

Reply:
Yes if not set there should be no health checks